scenario-compiler/src/main/java/pl/touk/nussknacker/engine/spel/internal/ConversionAndExtensionsAwareMethodInvoker.java

```diff
@@ -0,0 +1,32 @@
+package pl.touk.nussknacker.engine.spel.internal;
+
+import pl.touk.nussknacker.engine.extension.ExtensionMethods;
+import scala.PartialFunction;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+
+public class ConversionAndExtensionsAwareMethodInvoker {
+    private final ExtensionMethods extensionMethods;
+
+    public ConversionAndExtensionsAwareMethodInvoker(ExtensionMethods extensionMethods) {
+        this.extensionMethods = extensionMethods;
+    }
+
+    public Object invoke(Method method,
+                         Object target,
+                         Object[] arguments) throws IllegalAccessException, InvocationTargetException {
+        Class<?> methodDeclaringClass = method.getDeclaringClass();
+        if (target != null && target.getClass().isArray() && methodDeclaringClass.isAssignableFrom(List.class)) {
+            return method.invoke(RuntimeConversionHandler.convert(target), arguments);
```

Comment:
I understand that we want to apply runtime conversion to a collection-like object. 
But why we just do that in this line and pass the new target to the code below:

```scala
 PartialFunction<Class<?>, Object> extMethod =
            extensionMethods.invoke(method, target, arguments);
        if (extMethod.isDefinedAt(methodDeclaringClass)) {
            return extMethod.apply(methodDeclaringClass);
        } else {
            return method.invoke(target, arguments);
        }
```

which takes into consideration also extensions?

Reply:
Because of: https://github.com/TouK/nussknacker/pull/6875#discussion_r1778263570
But it can be simplified to:
```
        PartialFunction<Method, Object> extMethod = extensionMethods.invoke(target, arguments);
        if (extMethod.isDefinedAt(method)) {
            return extMethod.apply(method);
```

Reply:
no, no. I wanted to ask about sth else...
When the target is array-like we do:
```scala 
return method.invoke(RuntimeConversionHandler.convert(target), arguments);
```
(the conversion is applied and the method is invoked on the target).

Otherwise we do this:
```scala
 PartialFunction<Class<?>, Object> extMethod =
            extensionMethods.invoke(method, target, arguments);
        if (extMethod.isDefinedAt(methodDeclaringClass)) {
            return extMethod.apply(methodDeclaringClass);
        } else {
            return method.invoke(target, arguments);
        }
```
(which means: if the `methodDeclaringClass` has an extension, invoke it though the `extensionMethods` invoker, otherwise invoke the method directly).

So, my question is: what if we add some extension to a class that is array-like?
It looks like it won't be called because the first IF will handle this case.

Am I missing sth?

Reply:
Yes you are right, and I was planning to fix it with refactoring it to such implementation in which list methods are added as extension methods to array and in impl we converts array to list. I will add todo to highlight it I'm planning to refactor it in the future